flink-runtime/src/test/java/org/apache/flink/runtime/metrics/ReporterSetupTest.java

```diff
@@ -47,17 +50,30 @@
 import static org.junit.Assert.assertTrue;
 
 /** Tests for the {@link ReporterSetup}. */
-public class ReporterSetupTest extends TestLogger {
+@ExtendWith(TestLoggerExtension.class)
+class ReporterSetupTest {
+
+    @RegisterExtension
+    static final ContextClassLoaderExtension CONTEXT_CLASS_LOADER_EXTENSION =
+            ContextClassLoaderExtension.builder()
+                    .withServiceEntry(
+                            MetricReporterFactory.class,
+                            TestReporter1.class.getName(),
+                            TestReporter2.class.getName(),
+                            TestReporter11.class.getName(),
+                            TestReporter12.class.getName(),
+                            TestReporter13.class.getName())
+                    .build();
```

Comment:
What is the benefit of using a custom class loader here?

Reply:
In order for the service loader to work you need to list the implementations somewhere.
One option is adding it as service entry file under `src/test/META-INF/resources`, which is what most tests tend to use. The downside of this approach is that it this entry becomes part of the test-jar, and thus those entries can also be picked up in tests of downstream modules. This can lead to surprising behavior (e.g., something being loaded although we didn't expect it) or noisy logging messages (e.g., if you misconfigure a reporter we list all available implementations).
Essentially, this classloader extension provides better isolation between tests.